### ISSUE TITLE
Support middleware

### ISSUE THREAD

**Issue by mzabriskie** [rackt/react-router#824]
I was looking at @ryanflorence's [react-router-async-props](https://github.com/ryanflorence/react-router-async-props), and I really like the concept of being able to load async data from the server when a route is loaded. The downside to the current implementation is that you have to use a different `run`, and `RouteHandler` than is provided by react-router. So introducing something like this would require swapping out `Router.run` for react-router-async-props' `run` (not a big deal), and then replacing every `RouteHandler` with react-router-async-props` (less trivial depending on your app).

It would be ideal to have something more pluggable. Where we could just tell react-router to use some middleware that is handled internally, then the external API remains the same.

I am not proposing an API, just illustrating how it might work.

```js
var React = require('react');
var Router = require('react-router');
var { Route, RouteHandler, HistoryLocation } = Router;

var App = React.createClass({
  render () {
    return (
      <div>
        <h1>Hello World</h2>
        <RouteHandler/> {/* <-- actual Router.RouteHandler, nothing needs to change */}
      </div>
    );
  }
});

var FooHandler = React.createClass({
  statics: {
    asyncProps: { /* ... */ }
  },

  render () {
    return <div>foo</div>;
  }
});

var BarHandler = React.createClass({
  statics: {
    asyncProps: { /* ... */ }
  },

  render () {
    return <div>bar</div>;
  }
});

var routes = (
  <Route handler={App}>
    <Route name="foo" handler={FooHandler}/>
    <Route name="bar" handler={BarHandler}/>
  </Route>
);

// tell react-router to use react-router-async-props middleware
Router.use(require('react-router-async-props'));

// use the standard Router.run
Router.run(routes, HistoryLocation, (Handler, state, asyncProps) => {
  React.render(<Handler/>, document.body);
});
```

**Comment by ryanflorence** [comment]
Our current transition hooks could also just use this middleware, I async
handler loading too! #755

**Comment by mjackson** [comment]
My major hangup with building our own middleware system is that we start becoming more of a framework unto ourselves, instead of a library that you can easily use with some other system for composing middleware (of which there are quite a few, e.g. express or mach). Plus, I don't think there's enough value that we can create at this layer to justify our own middleware spec.

I think the fact that @ryanflorence was able to create an async props loading example on top of our current arch (using one of a few possible patterns) indicates that we're on the right track, insomuch as it is already possible (although maybe not incredibly convenient) to do so. I agree there is work to be done, but I guess I'd just like to move more in the direction of enabling various people's use cases instead of handing them a `use` function.

In other words, we do a few things and we do them well. When people show us what they *can't do* with what we've built, we can work to make that possible. I think that goes along with React's core arch as well (i.e. there are no React "plugins", just additions that build on top of the core).

Also, I'd really like to avoid a situation where we refactor everything to support middleware, then build about 3 of them and stop (see [superagent's "plugins"](https://github.com/visionmedia/superagent#plugins) for an example of what I'm talking about).

**Comment by ryanflorence** [comment]
1.0 `router.props.createElement` is basically a way to create "middleware"
